drv/transceivers-server/src/udp.rs

```diff
@@ -217,397 +215,588 @@ impl ServerImpl {
+        // iterate through the results and populate errors in the response buffer
+        // failure: whatever `HwError` specified by `failure_type`
+        // error: fpga communication issue
+        for module in modules.to_indices().map(LogicalPort) {
+            if module <= LogicalPortMask::MAX_PORT_INDEX {
+                if result.failure().is_set(module) {
+                    match failure_type {
+                        Some(e) => {
+                            let err_size =
+                                hubpack::serialize(&mut out[error_idx..], &e)
+                                    .unwrap();
+                            error_idx += err_size;
+                        }
+                        None => {
+                            ringbuf_entry!(Trace::UnexpectedFailure(
+                                ModuleId::from(module.as_mask())
+                            ));
+                            unreachable!();
+                        }
+                    }
+                } else if result.error().is_set(module) {
+                    let err_size = hubpack::serialize(
+                        &mut out[error_idx..],
+                        &HwError::FpgaError,
+                    )
+                    .unwrap();
+                    error_idx += err_size;
+                }
+            } else if requested_invalid_modules.is_set(module.0).unwrap() {
+                let err_size = hubpack::serialize(
+                    &mut out[error_idx..],
+                    &HwError::InvalidModuleIndex,
+                )
+                .unwrap();
+                error_idx += err_size;
+            }
+        }
+
+        // let the caller know how many error bytes we appended and which
+        // modules had problems
+        (
+            error_idx,
+            ModuleId(
+                requested_invalid_modules.0
+                    | result.error().0 as u64
+                    | result.failure().0 as u64,
+            ),
+        )
+    }
+
+    fn get_status(
+        &mut self,
+        modules: LogicalPortMask,
+        out: &mut [u8],
+    ) -> (usize, ModuleResult) {
+        // This will get the status of every module, so we will have to only
+        // select the data which was requested.
+        let (mod_status, full_result) = self.transceivers.get_module_status();
+        // adjust the result success mask to be only our requested modules
+        let desired_result = ModuleResult::new(
+            full_result.success() & modules,
+            full_result.failure(),
+            full_result.error(),
+        )
+        .unwrap();
+
+        // Write one bitfield per active port in the ModuleId which was
+        // successfully retrieved above.
         let mut count = 0;
-        for mask in modules.ports.to_indices().map(|i| 1 << i) {
+        for mask in modules
+            .to_indices()
+            .map(|p| p.as_mask())
+            .filter(|&m| !(desired_result.success() & m).is_empty())
```

Comment:
`desired_result.is_set(m)` perhaps?

Reply:
or even collapse the `map` statement, e.g.

```rust
        for mask in modules
            .to_indices()
            .filter(|p| desired_result.success().is_set(p.as_mask()))
```
(untested)

Reply:
Good idea! `p` is a `LogicalPort`, so we can just do `.is_set(p)`, no need for the `as_mask()`.